setup.py

```diff
@@ -1,68 +1,85 @@
-#
-if os.path.isdir('bin'):
-    setup_keywords['scripts'] = [fname for fname in glob.glob(os.path.join('bin', '*'))
-        if not os.path.basename(fname).endswith('.rst')]
-setup_keywords['provides'] = [setup_keywords['name']]
-setup_keywords['requires'] = ['Python (>2.7.0)']
-# setup_keywords['install_requires'] = ['Python (>2.7.0)']
-setup_keywords['zip_safe'] = False
-setup_keywords['use_2to3'] = False
-setup_keywords['packages'] = find_packages('py')
-setup_keywords['package_dir'] = {'':'py'}
-setup_keywords['cmdclass'] = {'module_file': DesiModule, 'version': DesiVersion, 'test': DesiTest, 'sdist': DistutilsSdist}
-setup_keywords['test_suite']='{name}.test.{name}_test_suite'.format(**setup_keywords)
-#
-# Autogenerate command-line scripts.
-#
-# setup_keywords['entry_points'] = {'console_scripts':['desiInstall = desiutil.install.main:main']}
-#
-# Run setup command.
-#
-setup(**setup_keywords)
+from setuptools import setup
+
+# First provide helpful messages if contributors try and run legacy commands
+# for tests or docs.
+
+API_HELP = """
+Note: Generating api.rst files is no longer done using 'python setup.py api'. Instead
+you will need to run:
+
+    desi_api_file
+
+which is part of the desiutil package. If you don't already have desiutil installed, you can install it with:
+
+    pip install desiutil
+"""
+
+MODULE_HELP = """
+Note: Generating Module files is no longer done using 'python setup.py api'. Instead
+you will need to run:
+
+    desiInstall
+
+or
+
+    desi_module_file
+
+depending on your exact situation.  desiInstall is preferred.  Both commands are
+part of the desiutil package. If you don't already have desiutil installed, you can install it with:
+
+    pip install desiutil
+"""
+
+VERSION_HELP = """
+Note: Generating version strings is no longer done using 'python setup.py version'. Instead
+you will need to run:
+
+    desi_set_version
+
+which is part of the desiutil package. If you don't already have desiutil installed, you can install it with:
+
+    pip install desiutil
+"""
+
+TEST_HELP = """
+Note: running tests is no longer done using 'python setup.py test'. Instead
+you will need to run:
+
+    pytest
+
+If you don't already have pytest installed, you can install it with:
+
+    pip install pytest
+"""
+
+DOCS_HELP = """
+Note: building the documentation is no longer done using
+'python setup.py {0}'. Instead you will need to run:
+
+    sphinx-build -W --keep-going -b html doc doc/_build/html
+
+If you don't already have Sphinx installed, you can install it with:
+
+    pip install Sphinx
+"""
+
+message = {'api': API_HELP,
+           'module_file': MODULE_HELP,
+           'test': TEST_HELP,
+           'version': VERSION_HELP,
+           'build_docs': DOCS_HELP.format('build_docs'),
+           'build_sphinx': DOCS_HELP.format('build_sphinx'), }
+
+for m in message:
+    if m in sys.argv:
+        print(message[m])
+        sys.exit(1)
```

Comment:
Thanks for including the documentation for these before/after cases.  Consider adding an "else" block to the for loop that prints a message if run with any option that isn't recognized, or otherwise using argparse to get an informative help message in that case.

That being said, this is a lot of boilerplate documentation that we probably don't want to copy into N>>1 other repos.  How about documenting it once in a web-friendly format (e.g. in the readthedocs for either desitemplate or desiutil), and then just have this code cite the URL of where to get help for any unrecognized option?

Reply:
I don't think that's a good idea, because we *don't know* what options may be used "under the hood" by pip or other installers.  We want everything *except* the specific cases in setup.py to be passed through.

It is reasonable to just pass through a URL in the message though.  That could refer to a page in desiutil that I'm already working on.

Reply:
I suggest replacing all the warning messages with a sentence containing this (prototype) URL: https://desiutil.readthedocs.io/en/module-script/helpers.html#replacing-setup-py